tests/mocha/variable_map_test.js

```diff
@@ -0,0 +1,251 @@
+      var varMapLength = this.variableMap.variableMap_[keys[0]].length;
+      assertEquals(1, varMapLength);
+
+      this.variableMap.createVariable('name1', 'type1');
+      assertVariableValues(this.variableMap, 'name1', 'type1', 'id1');
+      // Check that the size of the variableMap_ did not change.
+      keys = Object.keys(this.variableMap.variableMap_);
+      assertEquals(1, keys.length);
+      varMapLength = this.variableMap.variableMap_[keys[0]].length;
+      assertEquals(1, varMapLength);
+    });
+
+    test('Name already exists', function() {
+      // Expect that when a variable with the same name but a different type already
+      // exists, the new variable is created.
+      this.variableMap.createVariable('name1', 'type1', 'id1');
+
+      // Assert there is only one variable in the this.variableMap.
+      var keys = Object.keys(this.variableMap.variableMap_);
+      assertEquals(1, keys.length);
+      var varMapLength = this.variableMap.variableMap_[keys[0]].length;
+      assertEquals(1, varMapLength);
+
+      this.variableMap.createVariable('name1', 'type2', 'id2');
+      assertVariableValues(this.variableMap, 'name1', 'type1', 'id1');
+      assertVariableValues(this.variableMap, 'name1', 'type2', 'id2');
+      // Check that the size of the variableMap_ did change.
+      keys = Object.keys(this.variableMap.variableMap_);
+      assertEquals(2, keys.length);
+    });
+
+    test('Null type', function() {
+      this.variableMap.createVariable('name1', null, 'id1');
+      assertVariableValues(this.variableMap, 'name1', '', 'id1');
+    });
+
+    test('Undefined type', function() {
+      this.variableMap.createVariable('name2', undefined, 'id2');
+      assertVariableValues(this.variableMap, 'name2', '', 'id2');
+    });
+
+    test('Null id', function() {
+      sinon.stub(Blockly.utils, "genUid").returns('1');
+      this.variableMap.createVariable('name1', 'type1', null);
+      assertVariableValues(this.variableMap, 'name1', 'type1', '1');
+    });
+
+    test('Undefined id', function() {
+      sinon.stub(Blockly.utils, "genUid").returns('1');
+      this.variableMap.createVariable('name1', 'type1', undefined);
+      assertVariableValues(this.variableMap, 'name1', 'type1', '1');
+    });
+
+    test('Two variables same type', function() {
+      this.variableMap.createVariable('name1', 'type1', 'id1');
+      this.variableMap.createVariable('name2', 'type1', 'id2');
+
+      assertVariableValues(this.variableMap, 'name1', 'type1', 'id1');
+      assertVariableValues(this.variableMap, 'name2', 'type1', 'id2');
+    });
+
+    test('Two variables same name', function() {
+      // TODO(kozbial) shouldn't this fail?
+      this.variableMap.createVariable('name1', 'type1', 'id1');
+      this.variableMap.createVariable('name1', 'type2', 'id2');
+
+      assertVariableValues(this.variableMap, 'name1', 'type1', 'id1');
+      assertVariableValues(this.variableMap, 'name1', 'type2', 'id2');
```

Comment:
I'm confused that this works. Is it supposed to work?

Reply:
I'm thinking this is either:
A) Legacy behavior we want to test in case others depend on this change
B) Incorrect behavior we should have a bug for and link and skip this test (after adding an expect throws)

Reply:
I think this is intended, though likely implemented this way because of scratch.